libeos-payg/tests/clock-jump-source.c

```diff
@@ -0,0 +1,151 @@
+/* -*- mode: C; c-file-style: "gnu"; indent-tabs-mode: nil; -*-
+ *
+ * Copyright © 2019 Endless Mobile, Inc.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public License,
+ * v. 2.0. If a copy of the MPL was not distributed with this file, You can
+ * obtain one at https://mozilla.org/MPL/2.0/.
+ *
+ * Alternatively, the contents of this file may be used under the terms of the
+ * GNU Lesser General Public License Version 2.1 or later (the "LGPL"), in
+ * which case the provisions of the LGPL are applicable instead of those above.
+ * If you wish to allow use of your version of this file only under the terms
+ * of the LGPL, and not to allow others to use your version of this file under
+ * the terms of the MPL, indicate your decision by deleting the provisions
+ * above and replace them with the notice and other provisions required by the
+ * LGPL. If you do not delete the provisions above, a recipient may use your
+ * version of this file under the terms of either the MPL or the LGPL.
+ */
+
+#include "config.h"
+
+#include <glib.h>
+#include <libeos-payg/clock-jump-source.h>
+#include <locale.h>
+#include <unistd.h>
+
+/* 60 FPS should be enough for anyone. */
+#define ITERATIONS 30
+#define USEC_PER_MSEC 1000
+
+static gboolean
+timeout_cb (gpointer data)
+{
+  guint *called = (guint *) data;
+  *called += 1;
+  return G_SOURCE_CONTINUE;
+}
+
+/* Test the timeout firing once */
+static void
+test_once (void)
+{
+  g_autoptr(GError) local_error = NULL;
+  g_autoptr(GSource) source = NULL;
+  guint called = 0;
+  struct timespec original_ts;
+  struct timespec jump_ts;
+  int ret;
+
+  if (geteuid () != 0)
```

Comment:
I wonder if this test should also be marked "slow" or something else to prevent it being run by default – I think I would be surprised if building a package as root caused my clock to be tinkered with.

Reply:
Hmm, apparently systemd has a concept of an "unsafe" test but it's not a feature built into meson

Reply:
You can [tag tests with zero or more ‘suites’ in Meson](https://mesonbuild.com/Unit-tests.html), and then pass `--suite` or `--no-suite` to `meson test`. If you grep the GLib source for `suite`, for example, you’ll see how we use it. (Some tests are tagged as `slow` or `flaky`: https://gitlab.gnome.org/GNOME/glib/blob/master/gio/tests/meson.build#L76.)

Reply:
I think it'd be nice to do this but not a blocker.

I had a look at whether [libfaketime](https://github.com/wolfcw/libfaketime) supports timerfds (it doesn't) and whether time namespaces exist ([they don't, yet](https://lwn.net/Articles/766089/)). Boo.

Reply:
I added it to an "unsafe" suite, but it still runs by default. Perhaps that's good enough to give people a hint

Reply:
Why still run it by default?

Reply:
(sorry for the delay; didn't see this comment)

I added a commit to disable unsafe tests unless `-Dtests=unsafe` is used. I suppose it's fine that most people won't run it since autopkgtest will.

Reply:
> I suppose it's fine that most people won't run it since autopkgtest will.

[Nothing automatically runs these tests.](https://phabricator.endlessm.com/T18881)

Reply:
(I agree with not running these unsafe tests by default.)

Reply:
> Nothing automatically runs these tests.

(yet)

(patches welcome)